### DISCUSSION TITLE
build fails with wasm32-unknown-unknown

### DISCUSSION THREAD

**Discussion by gmesika-coti** [wasm-bindgen/wasm-bindgen#3513] [Q&A]
Hi

Trying to build a project that uses a crate, that has ursa as a dependency, which has openssl-sys

│ │ ├── openssl v0.10.55
│ │ │ ├── bitflags v1.3.2
│ │ │ ├── cfg-if v1.0.0
│ │ │ ├── foreign-types v0.3.2
│ │ │ │ └── foreign-types-shared v0.1.1
│ │ │ ├── libc v0.2.147
│ │ │ ├── once_cell v1.18.0
│ │ │ ├── openssl-macros v0.1.1 (proc-macro)
│ │ │ │ ├── proc-macro2 v1.0.63 ()
│ │ │ │ ├── quote v1.0.29 ()
│ │ │ │ └── syn v2.0.22 (*)
│ │ │ └── openssl-sys v0.9.90
│ │ │ └── libc v0.2.147

when running cargo build with --target=wasm32-unknown-unknown it fails on the error below
when running cargo build with --target=x86_64-unknown-linux-gnu its working fine (but of course does not generate what's expected)

what are my options:

1. edit the openssl-sys crate and add the code as suggested below , the use ... how do I build it locally and get other crates to use it?
2. run the build without compiling the package - can you suggest how? found few found posts but nothing that is working

similar issues: https://github.com/servo/rust-stb-image/issues/94

my error:

Compiling rand_chacha v0.3.1
Compiling time v0.1.45
Compiling zeroize_derive v1.4.2
error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./aes.rs:3:24
|
3 | pub const AES_ENCRYPT: c_int = 1;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./aes.rs:4:24
|
4 | pub const AES_DECRYPT: c_int = 0;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./aes.rs:6:22
|
6 | pub const AES_MAXNR: c_int = 14;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./aes.rs:7:27
|
7 | pub const AES_BLOCK_SIZE: c_int = 16;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./asn1.rs:6:23
|
6 | pub const V_ASN1_EOC: c_int = 0;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

error[E0412]: cannot find type c_int in this scope
--> /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.90/src/./asn1.rs:7:27
|
7 | pub const V_ASN1_BOOLEAN: c_int = 1;
| ^^^^^ not found in this scope
|
help: consider importing one of these items
|
1 | use core::ffi::c_int;
|
1 | use std::ffi::c_int;
|
1 | use std::os::raw::c_int;
|

**Comment by gmesika-coti** [comment]
clang details: 
/code # apk add clang
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz
(1/8) Installing libffi (3.4.4-r0)
(2/8) Installing xz-libs (5.2.9-r0)
(3/8) Installing libxml2 (2.10.4-r0)
(4/8) Installing zstd-libs (1.5.5-r0)
(5/8) Installing llvm15-libs (15.0.7-r0)
(6/8) Installing clang15-libs (15.0.7-r0)
(7/8) Installing libstdc++-dev (12.2.1_git20220924-r4)
(8/8) Installing clang15 (15.0.7-r0)

**Comment by daxpedda** [comment]
It looks like you are asking in the wrong place. This problem has nothing to do with `wasm-bindgen`.

In any case, you can't compile `openssl-sys` to `wasm32-unknown-unknown` and you can't fix that really.
What you can do, as you suggested, is avoid building it when you are using that target.
For that you can take a look at the [Cargo Book](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies).

**Comment by gmesika-coti** [comment]
Thanks for the reply!
I have looked extensively into the cargo book and tried to seek online help
couldn't find any real working solution on how not to rebuild all the dependencies 
If you have any idea what I could try more, appreciate!

**Reply by daxpedda** [reply]
> Thanks for the reply! I have looked extensively into the cargo book and tried to seek online help couldn't find any real working solution on how not to rebuild all the dependencies If you have any idea what I could try more, appreciate!

WDYM with not rebuilding all the dependencies? What are you trying to do that would require **re**building in the first place?

There's very little information to go on to help you here. It would be nice if you share some code and what exactly you are trying to achieve. The OP sounded like you were just trying to build something for the `wasm32-unknown-unknown` target. So it's not clear why are now trying to avoid **re**building all the dependencies and what you did or are trying to do to trigger that.

**Comment by gmesika-coti** [comment]
I didn't want to burden with too much information, trying to create a contract in wasm that includes hyperledger AnonCreds - https://github.com/hyperledger/anoncreds-rs - when it is included as a dependency in one of the contracts, it is trying to build it.
one of the dependencies of AnonCreds is ursa, which has openssl which has openssl-sys as a dependency.
that is not getting built with the error above 

thanks for the help!

**Reply by daxpedda** [reply]
If you take a look at [`ursa`s `Cargo.toml`](https://docs.rs/crate/ursa/0.3.7/source/Cargo.toml.orig), you can see that the dependency on `openssl` is already optional. You just have to not enable crate features that are using it.

On a quick glance the following default features already do that: `signatures_native` and `sharing_native`. But you are already [disabling default features](https://github.com/hyperledger/anoncreds-rs/blob/9c915bb77bc4e033cc6d28d45e330ee5bda26211/Cargo.toml#L39C1-L39C5), but you are enabling `cl_native`, which enables `openssl`.

Just to clarify, there is probably no sane way you can depend on `openssl` on the `wasm32-unknown-unknown` target, if it's not possible to avoid that, you will have to "fix" it in `ursa` yourself.

**Reply by daxpedda** [reply]
For reference: https://doc.rust-lang.org/cargo/reference/features.html.

**Comment by gmesika-coti** [comment]
I have pretty much saw (rust junior) what you mentioned, that features are disabled so the failures I have during build should not happen, but still they do.
how do you suggest I can fix ursa? ursa needs the types it takes from openssl-sys. how can that be changed?
why doesn't the cargo sees the libc crate during build of openssl-sys? and that is why im getting error like 
                  c_int;
    |          ^^^^^ not found in this scope

**Reply by daxpedda** [reply]
> I have pretty much saw (rust junior) what you mentioned, that features are disabled so the failures I have during build should not happen, but still they do.

As I said above, looking at https://github.com/hyperledger/anoncreds-rs `cl_native` is enabled, which in turn enables `openssl`. Again, it's hard to help if I can't take a look at what you are doing exactly.

> how do you suggest I can fix ursa? ursa needs the types it takes from openssl-sys. how can that be changed?

Whatever `ursa` is using `openssl` for, you would have to rewrite it to something compatible with Wasm.

> why doesn't the cargo sees the libc crate during build of openssl-sys? and that is why im getting error like
c_int;
| ^^^^^ not found in this scope

No clue. I don't think that's important anyway. You have bigger issues to fix first.

**Comment by gmesika-coti** [comment]
I will upload the project to github for reference soon, generally? trying to create a contract that uses HyperLedger AnonCreds as a dependency.

found this thread with similar issue: https://github.com/sfackler/rust-openssl/issues/1016

**Reply by daxpedda** [reply]
I didn't actually know that it's hypothetically possible to compile OpenSSL to Wasm already.
In any case, that would require more then just a brief look, you will have to figure this out on your own.

I'm happy to look at the crate feature problem you seem to have if you upload it to some repo, but don't expect anything more then that.

**Comment by gmesika-coti** [comment]
uploaded it to repo: https://github.com/gmesika-coti/cw-plus, the annoncreds as a dependency needs to be git cloned , and referenced in the toml , currently it is expected under /code/anoncreds-rs

**Reply by daxpedda** [reply]
Well `anoncreds-rs` needs to be fixed as I described above, did you do that? It would probably be helpful if you fork it and apply the fix and use it directly in `Cargo.toml` with a `git` property: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories.

You should try to respond in replies instead of creating a new "answer" every time.

**Reply by gmesika-coti** [reply]
aha, wow, yeah, missed that reply part 😂
I'm trying to fix the anoncreds-rs by not using the ursa which uses the openssl - it is a pretty long process
any idea on some other way?

**Reply by daxpedda** [reply]
I think the "easiest" way is to get `openssl` to compile to Wasm. But that's probably not easy at all.
Anything else will probably require some work to replace all OpenSSL usage with something that works on Wasm.
